Improve batch processing

al-serebrov opened the pull request:
* [cli_interface] Add hasbang to be able to make the file executable
* [es_loader] Add better way to split items in chunks, more verbose logging
* [docker-compose] Add volume to store loaded data (to keep it after container restart)

serebrov commented on es_loader/es_loader.py:
```diff
@@ -41,12 +72,49 @@ def _create_index(self):
 
+    def _get_job(self):
+        """Get job from ScrapingHubClient istance."""
+        self.job = self.sc.get_job(self.job_id)
```
This method is very small and seems to be only used once, so I would inline it (just use this code and remove the method). If you want extra clarity, renamve `sc` to `scrapinghub_client` so it will be very explicit - `self.scrapinhub_client.get_job(self.job_id)`.

serebrov commented on es_loader/es_loader.py:
```diff
@@ -41,12 +72,49 @@ def _create_index(self):
+
+    def _get_job_metadata(self):
+        """Get job metadata."""
+        self.metadata = self.job.metadata.list()
```
Same comment as above, for the _get_job method. Also `self.metadata` is only used in the `_get_scraped_items_count`, so maybe it's not worth putting it into `self`.

serebrov commented on es_loader/es_loader.py:
```diff
@@ -41,12 +72,49 @@ def _create_index(self):
+        and to calculate batches size"""
+        self._get_job_metadata()
+        for data in self.metadata:
+            if data[0] == 'scrapystats':
```
Do we need `else` here? What happens if we don't find `scrapystas` in data? Maybe just raise an exception at the end of the method if items_count is still not known.

serebrov commented on es_loader/es_loader.py:
```diff
@@ -41,12 +72,49 @@ def _create_index(self):
+        logger.debug(
+            'There are {} items scraped. Downloading...'.format(self.items_count)
+        )
+        items = self.job.items.iter()
```
I would also just inline this method for clarity.
When I am looking at `process_items`, it is impossible to immediately see what the item is in the `for item in self._get_items()`.
While if I see something like `for item in self.job.items.iter()`, I get a better idea of what is happening.

serebrov commented on es_loader/es_loader.py:
```diff
@@ -41,12 +72,49 @@ def _create_index(self):
+            item - an item from items() iterator of ScrapingHubClient.Job
+        """
+        if self.buffer_size == 0:
+            return
```
Does it make sense to have this check, considering that buffer_size is hardcoded in the constructor? 
Also, even if it weren't hardcoded, I'd better raise an exception instead of silently exiting and making the user wonder why there is nothing in the index.

serebrov commented on es_loader/es_loader.py:
```diff
@@ -56,14 +124,27 @@ def _index_item(self, item):
+            self.buffer_size = self.items_count - self.loaded_items_count
+            self.loaded_items_count += len(self.items_buffer)
+            if len(self.items_buffer) > 0:
+                self._send_items()
```
In general, I see two problems repeating in this PR: over-splitting (if I can say so) of the code into tiny methods and overusing `self`.

I think the logic that is implemented here belongs to the outer loop, where it can be handled in a more natural way, approximately like this:

```
    def process_items(self):
        """Entry point to the whole process.
        Calls all methods in order to process items of the job.
        """
        batch = []
        for item in self._get_items():

            index_action = {
                '_op_type': 'index',
                '_index': self.index_name,
                '_type': self.doc_type,
                'doc': dict(item)
            }

            batch.append(index_action)

            # if we accumulated enough items, send them in a bulk:
            if len(batch) >= self.buffer_size:
                self._bulk_send_items(batch)
                items = []

        # If there are any items left, send them in a bulk:
        if len(batch) > 0:
            self._bulk_send_items(batch)

    def _send_items(self, batch):
        self.loaded_items_count += len(batch)
        logger.debug('Bulk writing {} items'.format(len(batch)))
        helpers.bulk(client=self.es, actions=batch)
```

In the context of the loop we can simply check if some items are left after the loop with `if len(items) > 0`, while this is much more complex to do in the context of the one item processing (or inside the loop).

We also don't need to keep the items buffer inside `self`, it's OK to pass variables across class methods.

I usually keep to this pattern: everything related to the object state - variables attached to `self` - should be defined in the constructor, I am not doing something like `self.new_field = ...` inside the class methods.

al-serebrov commented:
Hi @serebrov !
Thank you so much for the review!
I've reworked ES loader according with your review, and also added some new functionality - now it calculates buffer size depending on a job items count, maybe it's not that optimal but I think it's better to have a hardcoded buffer size :)
Could you please have a look again?

serebrov commented on es_loader/es_loader.py:
```diff
@@ -72,79 +79,88 @@ def _create_index(self):
-        """Bulk write items to ElasticSearch."""
-        logger.debug('Bulk writing {} items'.format(len(self.items_buffer)))
-        helpers.bulk(client=self.es, actions=self.items_buffer)
+        buf = items_count // base_buffer_size
```
:open_mouth: I didn't know there is a `//` operator in python
